File: csharp/protein-translation/ProteinTranslation.cs
using System;
using System.Linq;

public static class ProteinTranslation
{
    public static string[] Proteins(string strand) =>
        strand
            .Chunk(3)
            .Select(chunk => string.Concat(chunk))
            .Select(ToProtein)
            .TakeWhile(protein => protein != "STOP")
            .ToArray();

    private static string ToProtein(string input) =>
        input switch
        {
            "AUG" => "Methionine",
            "UGG" => "Tryptophan",
            "UUU" or "UUC" => "Phenylalanine",
            "UUA" or "UUG" => "Leucine",
            "UAU" or "UAC" => "Tyrosine",
            "UGU" or "UGC" => "Cysteine",
            "UCU" or "UCC" or "UCA" or "UCG" => "Serine",
            "UAA" or "UAG" or "UGA" => "STOP",
            _ => throw new Exception("Invalid sequence")
        };
}

File: prolog/protein-translation/protein_translation.pl
:- use_module(library(dcg/basics)).

protein_sequence([]) --> [].
protein_sequence([]) --> stop, remainder(_).
protein_sequence([Protein|Proteins]) --> protein(Protein), protein_sequence(Proteins).

protein("Methionine") --> "AUG".
protein("Phenylalanine") --> "UUU" | "UUC".
protein("Leucine") --> "UUA" | "UUG".
protein("Serine") --> "UCU" | "UCC" | "UCA" | "UCG".
protein("Tyrosine") --> "UAU" | "UAC".
protein("Cysteine") --> "UGU" | "UGC".
protein("Tryptophan") --> "UGG".

stop --> "UAA" | "UAG" | "UGA".

translate(Rna, Proteins) :-
    string_codes(Rna, RnaCodes),
    phrase(protein_sequence(Proteins), RnaCodes), !.
